src/System.Globalization/tests/ISOWeekTests.cs

```diff
@@ -0,0 +1,215 @@
+            new DateData("2008-12-30", "2009-W01-2"),
+            new DateData("2008-12-31", "2009-W01-3"),
+            new DateData("2009-01-01", "2009-W01-4"),
+            new DateData("2009-12-31", "2009-W53-4"),
+            new DateData("2010-01-01", "2009-W53-5"),
+            new DateData("2010-01-02", "2009-W53-6"),
+            new DateData("2010-01-03", "2009-W53-7"),
+            new DateData("9999-12-31", "9999-W52-5"),
+        };
+
+        private static readonly YearData[] s_YearData =
+        {
+            new YearData(2005, "2005-01-03", "2006-01-01"),
+            new YearData(2006, "2006-01-02", "2006-12-31"),
+            new YearData(2007, "2007-01-01", "2007-12-30"),
+            new YearData(2008, "2007-12-31", "2008-12-28"),
+            new YearData(2009, "2008-12-29", "2010-01-03"),
+        };
+
+        public static IEnumerable<object[]> WeeksInYear_TestData()
+        {
+            for (int year = 0; year <= 400; year++)
+            {
+                bool isLong = s_YearsWith53Weeks.Contains(year);
+                yield return new object[] { 2000 + year, isLong };
+            }
+        }
+
+        public static IEnumerable<object[]> GetWeekOfYear_TestData()
+        {
+            return s_DateData.Select(x => new object[] {x.Date, x.Week});
+        }
+
+        public static IEnumerable<object[]> GetYear_TestData()
+        {
+            return s_DateData.Select(x => new object[] {x.Date, x.Year});
+        }
+
+        public static IEnumerable<object[]> ToDateTime_TestData()
+        {
+            return s_DateData.Select(x => new object[] {x.Year, x.Week, x.DayOfWeek, x.Date});
+        }
+
+        public static IEnumerable<object[]> GetYearStart_TestData()
+        {
+            return s_YearData.Select(x => new object[] {x.Year, x.StartDate});
+        }
+
+        public static IEnumerable<object[]> GetYearEnd_TestData()
+        {
+            return s_YearData.Select(x => new object[] {x.Year, x.EndDate});
+        }
+
+        [Theory, MemberData(nameof(GetWeekOfYear_TestData))]
+        public static void GetWeekOfYear(DateTime date, int expected)
+        {
+            Assert.Equal(expected, ISOWeek.GetWeekOfYear(date));
+        }
+
+        [Theory, MemberData(nameof(GetYear_TestData))]
+        public static void GetYear(DateTime date, int expected)
+        {
+            Assert.Equal(expected, ISOWeek.GetYear(date));
+        }
+
+        [Theory, MemberData(nameof(ToDateTime_TestData))]
+        public static void ToDateTime(int year, int week, DayOfWeek dayOfWeek, DateTime expected)
+        {
+            Assert.Equal(expected, ISOWeek.ToDateTime(year, week, dayOfWeek));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(10000)]
+        public static void ToDateTime_WithInvalidYear_Throws(int year)
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => ISOWeek.ToDateTime(year, 1, DayOfWeek.Friday));
```

Comment:
For all of these argument exceptions, is there a parameter name that can be validated as well?  Assuming yes, it'd be good to do that and to switch to using AssertExtensions.Throws, which is like Assert.Throws in that it has an overload that accepts a parameter name, but it's able to not enforce that name on .NET implementations that may throw away metadata.